bitcoin/src/address/mod.rs

```diff
@@ -662,6 +661,49 @@ impl Address<NetworkUnchecked> {
         }
     }
 
+    /// Checks that this is a mainnet address.
+    ///
+    /// For details about this mechanism, see section [*Parsing addresses*](Address#parsing-addresses)
+    /// on [`Address`].
+    #[inline]
+    pub fn require_mainnet(self) -> Result<Address, NetworkKindValidationError> {
+        self.require_network_kind(NetworkKind::Main)
+    }
+
+    /// Checks that this is not a mainnet address.
+    ///
+    /// For details about this mechanism, see section [*Parsing addresses*](Address#parsing-addresses)
+    /// on [`Address`].
+    #[inline]
+    pub fn require_not_mainnet(self) -> Result<Address, NetworkKindValidationError> {
+        self.require_network_kind(NetworkKind::Test)
+    }
+
+    /// Checks this address is valid for the `required` [`NetworkKind`].
+    ///
+    /// For details about this mechanism, see section [*Parsing addresses*](Address#parsing-addresses)
+    /// on [`Address`].
+    #[inline]
+    pub fn require_network_kind(self, required: NetworkKind) -> Result<Address, NetworkKindValidationError> {
```

Comment:
This would be ideal in theory but the issue is that we'd be the only software behaving this way, which would be confusing. So until a large chunk of the ecosystem is on board I'd rather not do this.